contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,821 @@
+`.Status.UpdateRevision` no update has occurred. 
+1. If the `Status.CurrentRevision` field is empty, then the StatefulSet has no 
+revision history. To initialize its revision history, the StatefulSet controller 
+will set both `.Status.CurrentRevision` and `.Status.UpdateRevision` to the 
+version of the current snapshot. 
+1. If the `.Status.CurrentRevision` is not empty, and if the 
+`.Status.UpdateRevision` is not equal to the version of the current snapshot, 
+the StatefulSet controller will set the `.Status.UpdateRevision` to the version 
+indicated by the current snapshot.
+
+### StatefulSet Revision History
+The StatefulSet controller will use the technique proposed in 
+[Controller History](https://github.com/kubernetes/community/pull/594) to 
+snapshot and version its target Object state.
+
+#### Snapshot Creation
+In order to snapshot a version of its target Object state, it will 
+serialize and store the `.Spec.Template` and `.Spec.VolumesClaimsTemplate` 
+along with the `.Generation` in each snapshot. Each snapshot will be labeled
+with the StatefulSet's `.Selector`.
+
+#### History Reconstruction
+As proposed in 
+[Controller History](https://github.com/kubernetes/community/pull/594), in 
+order to reconstruct the revision history of a StatefulSet, the StatefulSet 
+controller will select all snapshots based on its `Spec.Selector` and sort them 
+by the contained `.Generation`. This will produce an ordered set of 
+revisions to the StatefulSet's target Object state.
+
+#### History Maintenance 
+In order to prevent the revision history of the StatefulSet from exceeding 
+memory or storage limits, the StatefulSet controller will periodically prune 
+its revision history so that no more that `.Spec.RevisionHisotryLimit` non-live 
+versions of target Object state are preserved.
+
+### Update Completion
+The criteria for update completion is as follows.
+
+1. If the StatefulSet's `.Spec.UpdateStrategy.Type` is equal to 
+`OnDeleteStatefulSetStrategyType` then no version tracking is performed. It 
+this case an update can never be in progress.
+1. If the StatefulSet's `.Spec.UpdateStrategy.Type` is equal to 
+`PartitionedStatefulSetStrategyType` updates can not complete. The version 
+indicated `.Status.UpdateRevision` will only be applied to Pods with ordinals 
+in the sequence `(0,.Spec.UpdateStrategy.Partition.Ordinal]`
+1. If the StatefulSet's `.Spec.UpdateStrategy.Type` is equal to 
+`RollingUpdateStatefulSetStrategyType`, then an update is complete when the 
+StatefulSet is at its [target state](#target-state). The StatefulSet controller 
+will signal update completion as follows.
+    1. The controller will set `.Status.CurrentRevision` to the value of 
+    `.Staus.UpdateRevision`.
+    1. The controller will set `.Status.CurrentReplicas` to 
+    `.Status.UpdateReplicas`. Note that this value will be equal to 
+    `.Status.Replicas`.
+    1. The controller will set `.Status.UpdateReplicas` to 0.
```

Comment:
I've asked this before but I am still not sure why this is the case - why is `.Status.UpdateReplicas` reset to 0? All the replicas are updated at this point so this should equal `.Spec.Replicas`.